engine/access/state_stream/account_status_filter.go

```diff
@@ -0,0 +1,166 @@
+		// If `accountAddresses` exceeds the `DefaultAccountAddressesLimit`, it returns an error.
+		return AccountStatusFilter{}, fmt.Errorf("account limit exceeds, the limit is %d", DefaultMaxAccountAddresses)
+	}
+
+	// If `eventTypes` is empty, the filter returns all core events for any accounts.
+	if len(eventTypes) == 0 {
+		eventTypes = DefaultCoreEvents
+	}
+
+	// Creates an `EventFilter` with the provided `eventTypes`.
+	filter, err := NewEventFilter(config, chain, eventTypes, []string{}, []string{})
+	if err != nil {
+		return AccountStatusFilter{}, err
+	}
+
+	for _, address := range accountAddresses {
+		// Validate account address
+		addr := flow.HexToAddress(address)
+		if err := validateAddress(addr, chain); err != nil {
+			return AccountStatusFilter{}, err
+		}
+
+		// If there are non-core event types at this stage, it returns an error from `addCoreEventFieldFilter`.
+		for eventType := range filter.EventTypes {
+			// use the hex with prefix address to make sure it will match the cadence address
+			err = filter.addCoreEventFieldFilter(eventType, addr.HexWithPrefix(), chain)
+			if err != nil {
+				return AccountStatusFilter{}, err
+			}
+		}
+	}
+
+	return AccountStatusFilter{
+		EventFilter: &filter,
+	}, nil
+}
+
+// GroupCoreEventsByAccountAddress extracts account-related core events from the provided list of events.
+// It filters events based on the account field specified by the event type and organizes them by account address.
+// Parameters:
+// - events: The list of events to extract account-related core events from.
+// - log: The logger to log errors encountered during event decoding and processing.
+// Returns:
+//   - A map[string]flow.EventsList: A map where the key is the account address and the value is a list of
+//     account-related core events associated with that address.
+func (f *AccountStatusFilter) GroupCoreEventsByAccountAddress(events flow.EventsList, log zerolog.Logger) map[string]flow.EventsList {
+	allAccountProtocolEvents := make(map[string]flow.EventsList)
+
+	for _, event := range events {
+		fields, fieldValues, err := getEventFields(&event)
+		if err != nil {
+			log.Info().Err(err).Msg("could not get event fields")
+			continue
+		}
+
+		accountField := f.EventFieldFilters[event.Type]
+		for i, field := range fields {
+			_, ok := accountField[field.Identifier]
+			if ok {
+				address := fieldValues[i].String()
+				allAccountProtocolEvents[address] = append(allAccountProtocolEvents[address], event)
+			}
+		}
+	}
+
+	return allAccountProtocolEvents
+}
+
+// validateCoreEventTypes validates the provided event types against the default core event types.
+// It returns an error if any of the provided event types are not in the default core event types list. Note, an empty
+// event types array is also valid.
+func validateCoreEventTypes(eventTypes []string) error {
+	for _, eventType := range eventTypes {
+		isMatch := false
+		for _, coreEventType := range DefaultCoreEvents {
```

Comment:
we should avoid doing an `N*M`  iteration here. Instead, how about creating a map and using that for the lookups?
```golang

var defaultCoreEventsMap map[string]struct{}

func init() {
	defaultCoreEventsMap = make(map[string]struct{}, len(DefaultCoreEvents))
	for i := range DefaultCoreEvents {
		defaultCoreEventsMap[DefaultCoreEvents[i]] = struct{}{}
	}
}